[KubeRay][Doc] KubeRay doc migration (1/n) (#38647)
ray-project/ray: doc/source/cluster/kubernetes/k8s-ecosystem/prometheus-grafana.md

(kuberay-prometheus-grafana)=

> # Prometheus & Grafana

Reviewer: ```suggestion
# Using Prometheus and Grafana
```

This section will describe how to monitor Ray Clusters in Kubernetes using Prometheus & Grafana.

> If you do not have any experience with Prometheus and Grafana on Kubernetes, I strongly recommend you watch this [YouTube playlist](https://youtube.com/playlist?list=PLy7NrYWoggjxCF3av5JKwyG7FFF9eLeL4).

Reviewer: ```suggestion
If you do not have any experience with Prometheus and Grafana on Kubernetes, watch this [YouTube playlist](https://youtube.com/playlist?list=PLy7NrYWoggjxCF3av5JKwyG7FFF9eLeL4).
```

## Preparation

> Please clone the [KubeRay repository](https://github.com/ray-project/kuberay) and checkout the `master` branch.

Reviewer: ```suggestion
Clone the [KubeRay repository](https://github.com/ray-project/kuberay) and checkout the `master` branch.
```

This tutorial requires several files in the repository.

> ## Step 1: Create a Kubernetes cluster with Kind.

Reviewer: ```suggestion
## Step 1: Create a Kubernetes cluster with Kind.
```
```suggestion
## Step 1: Create a Kubernetes cluster with Kind
```

```sh
# path: ray-operator/config/samples/
kubectl apply -f ray-cluster.embed-grafana.yaml

# Check ${RAYCLUSTER_HEAD_POD}
kubectl get pod -l ray.io/node-type=head

# Example output:
# NAME                            READY   STATUS    RESTARTS   AGE
# raycluster-kuberay-head-btwc2   1/1     Running   0          63s

# Wait until all Ray Pods are running and forward the port of the Prometheus metrics endpoint in a new terminal.
kubectl port-forward --address 0.0.0.0 ${RAYCLUSTER_HEAD_POD} 8080:8080
curl localhost:8080

# Example output (Prometheus metrics format):
# # HELP ray_spill_manager_request_total Number of {spill, restore} requests.
# # TYPE ray_spill_manager_request_total gauge
# ray_spill_manager_request_total{Component="raylet",NodeAddress="10.244.0.13",Type="Restored",Version="2.0.0"} 0.0

# Ensure that the port (8080) for the metrics endpoint is also defined in the head's Kubernetes service.
kubectl get service

# NAME                          TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)                                         AGE
# raycluster-kuberay-head-svc   ClusterIP   10.96.201.142   <none>        6379/TCP,8265/TCP,8080/TCP,8000/TCP,10001/TCP   106m
```

> * KubeRay will expose a Prometheus metrics endpoint in port **8080** via a built-in exporter by default. Hence, we do not need to install any external exporter.

Reviewer: ```suggestion
* KubeRay exposes a Prometheus metrics endpoint in port **8080** via a built-in exporter by default. Hence, we do not need to install any external exporter.
```
